### PR TITLE
Add Differential Diffusion to Kolors

### PR DESCRIPTION
**Pull request by saqlain2204**
# What does this PR do?

Adds [Differential Diffusion](https://arxiv.org/abs/2306.00950) to Kolors

Fixes Partially https://github.com/huggingface/diffusers/issues/8924(Kolors)

@a-r-r-o-w @DN6 

```py
import torch

from diffusers.utils import load_image
from pipeline_kolors_differential_img2img import KolorsDifferentialImg2ImgPipeline

pipe = KolorsDifferentialImg2ImgPipeline.from_pretrained(
             "Kwai-Kolors/Kolors-diffusers", variant="fp16", torch_dtype=torch.float16
       )
pipe.enable_model_cpu_offload()

source_image = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/differential/20240329211129_4024911930.png"
)
map = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/differential/gradient_mask_2.png"
)
prompt = "a green pear"
negative_prompt = "blurry"

image = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    image=source_image,
    num_inference_steps=28,
    guidance_scale=4.5,
    strength=1.0,
    map=map,
).images[0]
```

### CODE REVIEW

**Comment by saqlain2204** [comment]
I am not able to lead kolors in colab free tier. Tried to resolve it by using memory optimizations, but it dint work. Am I going wrong somewhere?

**Comment by a-r-r-o-w** [comment]
Thanks for working on this! For Kolors, I think @asomoza might be able to share some suggestions on how to save memory and minimal VRAM requirements as he worked on it. If it doesn't work on a free-tier colab, I'll try and test the PR :)

**Comment by saqlain2204** [comment]
Sure. Thanks :)

**Comment by asomoza** [comment]
Hi @saqlain2204 sorry I missed this for so long. Kolors uses a big text encoder and a custom one that isn't in transformers yet so we can't optimize it for now, so for the time being, we can't use a quantized version of it.

As I remember the text encoder is around 10GB and the unet 6 GB so they won't fit in the free tier colab GPU, don't remember how much RAM it has to store them in the CPU, but it should work with `enable_model_cpu_offload`. There was a problem before where the VRAM wasn't getting freed but that was fixed a while ago, maybe you can try again.

If it doesn't work like this, sadly we won't be able to optimize it just yet so it can run on the free tier colab.

**Comment by yiyixuxu** [comment]
should we merge this?

**Comment by saqlain2204** [comment]
Hi @asomoza ,
Thanks for the suggestions. Will try running it by making the above changes and update it here.

**Comment by asomoza** [comment]
@saqlain2204 can you update the description with the typical examples with use for diff-diff, you can get them from the [SD3](https://github.com/huggingface/diffusers/pull/8679) implementation.

This way you and we can make sure it works which is pretty much the only requirement for community pipelines. From a quick look, it seems you're missing the map/mask input so there's no way to pass the depthmap/mask to the pipeline call method.

**Comment by saqlain2204** [comment]
I have updated. Can you please review it.

File: examples/community/pipeline_kolors_differential_img2img.py [hunk 1]
```diff
@@ -0,0 +1,1288 @@
...
+                batch_size * num_images_per_prompt,
+                num_channels_latents,
+                height,
+                width,
+                init_image,
+                latent_timestep,
+                prompt_embeds.dtype,
+                device,
+                generator,
+            )
```
Comment (asomoza) [inline examples/community/pipeline_kolors_differential_img2img.py:1083]:
```suggestion
        if latents is None:
            latents = self.prepare_latents(
                init_image,
                latent_timestep,
                batch_size,
                num_images_per_prompt,
                prompt_embeds.dtype,
                device,
                generator,
                add_noise,
            )
```

File: examples/community/pipeline_kolors_differential_img2img.py [hunk 2]
```diff
@@ -0,0 +1,1288 @@
...
+            batch_size * num_images_per_prompt,
+            num_channels_latents,
+            height,
+            width,
+            init_image,
+            timesteps,
+            prompt_embeds.dtype,
+            device,
+            generator,
+        )
```
Comment (asomoza) [inline examples/community/pipeline_kolors_differential_img2img.py:1149]:
```suggestion
        original_with_noise = self.prepare_latents(
            init_image, timesteps, batch_size, num_images_per_prompt, prompt_embeds.dtype, device, generator
        )
```

**Review by asomoza — CHANGES_REQUESTED** [review]
we need a couple of fixes, the rest looks good

**Comment by asomoza** [comment]
@saqlain2204 can you please run `make style` and `make quality`

**Comment by asomoza** [comment]
thanks!

**Comment by saqlain2204** [comment]
👍

**Comment by saqlain2204** [comment]
Thanks @a-r-r-o-w @asomoza !
